pkg/cache/memory/kvcache.go

```diff
@@ -43,102 +40,92 @@ func NewKeyValueCache[T any](opts ...KeyValueCacheOptionFn) (*keyValueCache[T],
 	}
 
 	return &keyValueCache[T]{
-		values: make(map[string]cacheValue[T]),
+		values: xsync.NewMap[string, cacheValue[T]](),
 		config: config,
 	}, nil
 }
 
 // Get retrieves the value from the cache with the given key.
 func (c *keyValueCache[T]) Get(key string) (T, bool) {
 	var zero T
-	c.valuesMu.RLock()
-	defer c.valuesMu.RUnlock()
-
-	cachedValue, exists := c.values[key]
-	if !exists {
+	v, ok := c.values.Load(key)
+	if !ok {
 		return zero, false
 	}
-
-	isCacheValueExpired := time.Since(cachedValue.cachedAt) > c.config.ttl
-	if isCacheValueExpired {
-		// DEV_NOTE: Not pruning here to optimize concurrent speed:
-		// - Read lock alone would be insufficient for pruning
-		// - Next Set() call will overwrite the value
-		// - If values aren't subsequently set, maxKeys config will eventually trigger
-		//   pruning of TTL-expired values
+	if time.Since(v.cachedAt) > c.config.ttl {
+		// Opportunistic prune (no need for atomic compute here).
+		c.values.Delete(key)
 		return zero, false
 	}
-
-	return cachedValue.value, true
+	return v.value, true
 }
 
 // Set adds or updates the value in the cache for the given key.
 func (c *keyValueCache[T]) Set(key string, value T) {
-	c.valuesMu.Lock()
-	defer c.valuesMu.Unlock()
+	c.values.Store(key, cacheValue[T]{value: value, cachedAt: time.Now()})
 
-	c.values[key] = cacheValue[T]{
-		value:    value,
-		cachedAt: time.Now(),
+	if c.config.maxKeys > 0 && int64(c.values.Size()) > c.config.maxKeys {
```

Comment:
Why do we need to check for `maxKeys > 0`?

Not worried about the performance, just confused.
- Could (should?) we remove it?
- Do we need to #PUC?